### PR TITLE
Example threejs terrain

### PR DESCRIPTION
**Pull request by MichaelLangbein**
## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - N/A Include before/after visuals or gifs if this PR includes visual changes.
 - N/A Write tests for all new functionality.
 - N/A Document any changes to public APIs.
 - N/A Post benchmark scores.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

## Description
- Adds documentation: new example of using a three.js `CustomLayer` together with maplibre's terrain.
- Relates to [this issue](https://github.com/maplibre/maplibre-gl-js/issues/3427).
- Places a threejs scene on terrain by accounting for maplibre per default displaying all objects at the height of `center` - even if the objects themselves are located at a point on the terrain that lies higher or lower than `center`.
- It does so by simply translating the scene by `sceneElevation - terrainCenterElevation`.

Needs some review:
1. Is the `sceneElevation - terrainCenterElevation` approach suggested here really idiomatic?
2. Couldn't run `generate-images` as described [here](https://github.com/maplibre/maplibre-gl-js/blob/8edbac09725f807aae91009d31c30c3d5168b066/docs/README.md) - is the provided image sufficient?

### CODE REVIEW

**Review by HarelM — APPROVED** [review]
Overall this looks good, I'm not familiar with three to fully understand this, but I think it's OK.

**Comment by MichaelLangbein** [comment]
> The three.js part itself looks good. There is a lot more math here than in the old debug page I posted in the issue thread, but I do not know enough about it to know if this is all needed. @MichaelLangbein, do you think you could you take a look at the [old debug page](https://github.com/maplibre/maplibre-gl-js/blob/7bab10e8dfd1680471126b02b4849423a5f50365/test/debug-pages/terrain-3d-object.html) and help us understand the differences?

Sure thing, will do! I'll add some explaining comments some time after tonight. I can already say that there's not much change relative to the Debug-Page. 

- I've attempted to make things a little simpler by removing a little of the rotation logic.
- I'm using a scene with two models in it to prove that the process works with objects at different elevations.
- That helper function `calculateDistanceMercatorToMeters` is just a quick-and-dirty way to get distances in threejs-scene coordinates from Mercator coordinates. It won't work if objects are far apart, which I will add as a comment.

The whole rest of the matrix math stems from the existing example in the docs, but I'll add some comments there, too.

**Comment by HarelM** [comment]
Not sure I know how to answer this question, can you clarify it a bit?

File: test/examples/add-3d-model-with-terrain.html [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+                        .scale(new THREE.Vector3(sceneTransform.scale, -sceneTransform.scale, sceneTransform.scale));
+
+                    this.camera.projectionMatrix = m.multiply(l);
+                    this.renderer.resetState();
+                    this.renderer.render(this.scene, this.camera);
+                    map.triggerRepaint();
+                }
+            };
+
+            map.on('load', () => {
```
[the remark this answers is not in the document]
Comment (MichaelLangbein) [reply test/examples/add-3d-model-with-terrain.html]: Ah, didn't know about that. Has been changed - thanks!

**Comment by MichaelLangbein** [comment]
> Not sure I know how to answer this question, can you clarify it a bit?

This is my understanding of how the variables in `map.projection` are to be interpreted - it differs a little bit from the graphic shown [here](https://github.com/maplibre/maplibre-gl-js/issues/1318#issuecomment-1167793345) 

drawing

**Comment by MichaelLangbein** [comment]
Alright, I've switched the code to use `map.queryTerrainElevation`, which did save me a few lines of calcultions.
I've also added a bunch of comments explaining:
- the reasoning behind how the threejs-scene has been placed and transformed (which I think is now a little bit nicer than on my first attempt, because I'm now transforming the whole scene instead of each object individually)
- a few words about what `queryTerrainElevation` actually returns, particularly that it already returns the difference with the elevation of `center`.

**Comment by MichaelLangbein** [comment]
Hmm, sometimes the unit tests don't seem to pass. I haven't changed anything about maplibre itself - is this an actual error or just a CI issue?

**Comment by neodescis** [comment]
Thank you @MichaelLangbein!

**Pull request merged**